File: BinaryTreeLevelOrderTraversal.py
"""
102. Binary Tree Level Order Traversal
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]
"""
# O(n)
# O(1)

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        cur, res = [root], []
        while cur:
            next, val = [], []
            for node in cur:
                val.append(node.val)
                if node.left:
                    next.append(node.left)
                if node.right:
                    next.append(node.right)
            res.append(val)
            cur = next
        return res

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right.left = TreeNode(5)
    print(Solution().levelOrder(root))

File: Java/BinaryTreeLevelOrderTraversal.java
import java.util.*;
public class BinaryTreeLevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> vals;
        List<TreeNode> cur = new ArrayList<TreeNode>();
        List<TreeNode> next;
        if(root == null) {return res;}
        cur.add(root);
        while(cur.size() != 0) {
            vals = new ArrayList<Integer>();
            next = new ArrayList<TreeNode>();
            for(TreeNode node: cur) {
                vals.add(node.val);
                if(node.left != null) {next.add(node.left);}
                if(node.right != null) {next.add(node.right);}
            }
            res.add(vals);
            cur = next;
        }
        return res;
    }

}
